README.md
<a href="http://www.bioconductor.org/packages/release/bioc/html/brainflowprobes.html#since"><img border="0" src="http://www.bioconductor.org/shields/years-in-bioc/brainflowprobes.svg" title="How long since the package was first in a released Bioconductor version (or is it in devel only)."></a> <a href="https://bioconductor.org/packages/stats/bioc/brainflowprobes/"><img border="0" src="http://www.bioconductor.org/shields/downloads/brainflowprobes.svg" title="Percentile (top 5/20/50% or 'available') of downloads over last 6 full months. Comparison is done across all package categories (software, annotation, experiment)."></a> <a href="https://support.bioconductor.org/t/brainflowprobes/"><img border="0" src="http://www.bioconductor.org/shields/posts/brainflowprobes.svg" title="Support site activity, last 6 months: tagged questions/avg. answers per question/avg. comments per question/accepted answers, or 0 if no tagged posts."></a> <a href="http://www.bioconductor.org/packages/release/bioc/html/brainflowprobes.html#svn_source"><img border="0" src="http://www.bioconductor.org/shields/commits/bioc/brainflowprobes.svg" title="average Subversion commits (to the devel branch) per month for the last 6 months"></a>

Bioc-release <a href="http://www.bioconductor.org/packages/release/bioc/html/brainflowprobes.html#archives"><img border="0" src="http://www.bioconductor.org/shields/availability/release/brainflowprobes.svg" title="Whether the package is available on all platforms; click for details."></a> <a href="http://bioconductor.org/checkResults/release/bioc-LATEST/brainflowprobes/"><img border="0" src="http://www.bioconductor.org/shields/build/release/bioc/brainflowprobes.svg" title="build results; click for full report"></a>,
Bioc-devel <a href="http://www.bioconductor.org/packages/devel/bioc/html/brainflowprobes.html#archives"><img border="0" src="http://www.bioconductor.org/shields/availability/devel/brainflowprobes.svg" title="Whether the package is available on all platforms; click for details."></a> <a href="http://bioconductor.org/checkResults/devel/bioc-LATEST/brainflowprobes/"><img border="0" src="http://www.bioconductor.org/shields/build/devel/bioc/brainflowprobes.svg" title="build results; click for full report"></a>.

Bioc-release <a href="https://bioconductor.org/developers/how-to/unitTesting-guidelines/#coverage"><img border="0" src="http://www.bioconductor.org/shields/coverage/release/brainflowprobes.svg" title="Test coverage percentage, or 'unknown'"></a>, Bioc-devel <a href="https://codecov.io/github/Bioconductor-mirror/brainflowprobes?branch=master"><img border="0" src="http://www.bioconductor.org/shields/coverage/devel/brainflowprobes.svg" title="Test coverage percentage, or 'unknown'"></a>

brainflowprobes
===============

This package can be used to annotate candidate target probe sequences designed to label RNA in nuclei isolated from human postmortem brain for flow cytometry. This package also includes functions that visualize candidate sequence expression in four postmortem brain datasets that address nuclear expression, effects of RNA degradation and flow cytometric sorting, and cell type-specific expression to increase the likelihood of success as a custom target probe. Please refer to the BrainFlow publication or the [PrimeFlow RNA(TM) Assay Kit](https://www.thermofisher.com/order/catalog/product/88-18005-204) literature for more information.

# Installation instructions

Get R 3.6.x from [CRAN](http://cran.r-project.org/).

```R
## From Bioconductor
install.packages("BiocManager")
BiocManager::install('brainflowprobes')

## Currently: BiocManager::install('LieberInstitute/brainflowprobes')
```

# Vignettes

The vignettes for this package can be viewed via [Bioconductor's website](http://www.bioconductor.org/packages/brainflowprobes).



NEWS
CHANGES IN VERSION 0.99.3
-------------------------

NEW FEATURES

    o Created three utility functions: check_pdf(), get_nearest_annotation()
       and get_region_cov() and resolved other tasks suggested by
       @Liubuntu at https://github.com/Bioconductor/Contributions/issues/1191#issuecomment-522762540
       during the Bioconductor submission process.

CHANGES IN VERSION 0.99.0
-------------------------

NEW FEATURES

    o Created the four_panels(), region_info() and plot_coverage() functions.
    o My first R/Bioconductor package! ^^


tests/testthat.R
library(testthat)
library(brainflowprobes)

test_check("brainflowprobes")


tests/testthat/test-pdf_check.R
## Create some fake pdf files
tmp_dir <- tempdir()
pdf(file.path(tmp_dir, 'four_panels.pdf'))
plot(1, 1)
dev.off()
pdf(file.path(tmp_dir, 'regionCoverage_fractionedData.pdf'))
plot(1, 1)
dev.off()

test_that('pdfs are not overwritten', {
    expect_error(four_panels("chr20:10286777-10288069:+",
        PDF = file.path(tmp_dir, 'four_panels.pdf')))
    expect_error(four_panels("chr20:10286777-10288069:+",
        PDF = file.path(tmp_dir, 'four_panels')))
    expect_error(plot_coverage("chr20:10286777-10288069:+",
        PDF = file.path(tmp_dir, 'regionCoverage_fractionedData.pdf')))
    expect_error(plot_coverage("chr20:10286777-10288069:+",
        PDF = file.path(tmp_dir, 'regionCoverage_fractionedData')))
})


## Choose a random PDF file
PDF <- file.path(tempdir(),
    paste0('test_', stats::runif(1, max = 1e10))
)
PDF

## Initially this works because the output PDF does not exist.
## It also adds the PDF extension if the user didn't supply it.
test_that('check_pdf part own', {
    expect_identical(check_pdf(PDF), paste0(PDF, '.pdf'))
})
PDF <- check_pdf(PDF)

## Create a dummy PDF file
pdf(file = PDF)
plot(1, 1)
dev.off()

## Now it doesn't work since the PDF file already exists.
test_that('check_pdf part two', {
    expect_error(check_pdf(PDF), 'already exists! Rename or erase it')
})


tests/testthat/test-get_region_cov.R
test_that("get_region_cov", {
    expect_equal(
        get_region_cov(COVERAGE = four_panels_example_cov),
        four_panels_example_cov
    )
    ## Order of the names doesn't matter as long as all the data is there
    expect_equal(
        get_region_cov(
            COVERAGE = four_panels_example_cov[
                rev(names(brainflowprobes::pd))]),
        four_panels_example_cov
    )
    expect_error(
        get_region_cov(COVERAGE = list('hola' = 1))
    )
    expect_error(
        get_region_cov(
            COVERAGE = list(
                'Sep' = matrix(),
                'Deg' = matrix(),
                'Cell' = matrix(),
                'Sort' = matrix()
            )
        ),
        'region coverage data.frames'
    )
    expect_error(
        get_region_cov(
            COVERAGE = list(
                'Sep' = list('1' = matrix()),
                'Deg' = list('1' = matrix()),
                'Cell' = list('1' = matrix()),
                'Sort' = list('1' = matrix())
            )
        ),
        'region coverage data.frames'
    )
    expect_error(
        get_region_cov(
            COVERAGE = list(
                'Sep' = list('1' = data.frame()),
                'Deg' = list('1' = data.frame()),
                'Cell' = list('1' = data.frame()),
                'Sort' = list('1' = data.frame())
            )
        ),
        'should have a column per each of the samples in'
    )
    expect_error(
        get_region_cov(
            COVERAGE = list(
                'Sep' = list('1' = data.frame(matrix(ncol = 23, nrow = 2))),
                'Deg' = list('1' = data.frame(matrix(ncol = 40))),
                'Cell' = list('1' = data.frame(matrix(ncol = 466))),
                'Sort' = list('1' = data.frame(matrix(ncol = 12)))
            )
        ),
        'should have the same number of rows'
    )
})


R/utils.R
#' Check the PDF file
#'
#' This function checks that the PDF file does not exist to avoid
#' overwriting a plot the user has previously made.
#'
#' This is an utility function used by \link{four_panels} and
#' \link{plot_coverage}.
#'
#' @inheritParams four_panels
#'
#' @author Leonardo Collado-Torres
#' @return The path to the PDF file if the file doesn't exist. It appends
#' the pdf file extension if it was absent from `PDF`.
#' @export
#' @examples
#'
#' ## Choose a random PDF file
#' PDF <- file.path(tempdir(),
#'     paste0('test_', stats::runif(1, max = 1e10))
#' )
#' PDF
#'
#' ## Initially this works because the output PDF does not exist.
#' PDF <- check_pdf(PDF)
#' ## It also adds the PDF extension if the user didn't supply it.
#' PDF
#'
#' ## Create a dummy PDF file
#' pdf(file = PDF)
#' plot(1, 1)
#' dev.off()
#'
#' ## Now it doesn't work since the PDF file already exists.
#' testthat::expect_error(check_pdf(PDF), 'already exists! Rename or erase it')
#'

check_pdf <- function(PDF = 'four_panels.pdf') {
    pdf_file <- PDF
    if (!grepl("pdf$", tolower(PDF)))
        pdf_file <- paste0(PDF, ".pdf")
    if (file.exists(pdf_file))
        stop(paste("The file",
            pdf_file,
            "\nalready exists! Rename or erase it before proceeding."),
            call. = FALSE)
    return(pdf_file)
}

#' Compute the nearest annotation to the genes in brainflowprobes
#'
#' For a given set of genomic regions, this function computes the nearest
#' annotation information using the \link{genes} object distributed in this
#' package.
#'
#' This is an utility function used by \link{region_info}, \link{four_panels}
#' and \link{plot_coverage}.
#'
#' @param gr A [GenomicRanges::GRanges()] object.
#' @inheritParams four_panels
#'
#' @return The [bumphunter::matchGenes()] output for the annotation information
#' using the `genes` object in this package (subset to only the coding elements
#' if `CODING_ONLY` was set to `TRUE`).
#' @export
#' @author Leonardo Collado-Torres
#' @examples
#'
#' gr <- GenomicRanges::GRanges('chr10:135379301-135379311:+')
#'
#' get_nearest_annotation(gr)
#' get_nearest_annotation(gr, CODING_ONLY = TRUE)
#'
get_nearest_annotation <- function(gr, CODING_ONLY = FALSE) {
    gr_subject <- if(CODING_ONLY) {
        brainflowprobes::genes[!is.na(brainflowprobes::genes$CSS)]
    } else {
        brainflowprobes::genes
    }
    nearestAnnotation <- bumphunter::matchGenes(x = gr, subject = gr_subject)
    return(nearestAnnotation)
}


#' Check or compute the region coverage from the datasets in brainflowprobes
#'
#' This utility function checks the user-provided region data.frame coverage
#' list (`COVERAGE`) or computes a new one using \link{brainflowprobes_cov}.
#' This is used by \link{four_panels} and \link{plot_coverage}.
#'
#' @inheritParams four_panels
#' @inheritParams brainflowprobes_cov
#'
#' @return If `COVERAGE` is provided and all checks pass, then this function
#' returns `COVERAGE`. Otherwise, it computes a new region coverage data.frame
#' list using \link{brainflowprobes_cov}.
#'
#' @export
#' @author Leonardo Collado-Torres
#' @examples
#'
#' ## If all checks pass, then it returns the COVERAGE
#' stopifnot(identical(
#'     get_region_cov(COVERAGE = four_panels_example_cov),
#'     four_panels_example_cov
#' ))
#'
get_region_cov <- function(REGION, COVERAGE = NULL, VERBOSE = TRUE, PD = brainflowprobes::pd) {
    if(is.null(COVERAGE)) {
        regionCov <- brainflowprobes_cov(
            REGION = REGION,
            PD = PD,
            VERBOSE = VERBOSE
        )
    } else {
        stopifnot(is.list(COVERAGE))
        if(!all(c('Sep', 'Deg', 'Cell', 'Sort') %in% names(COVERAGE))) {
            stop("'COVERAGE' should be a list with the elements:\n'",
                paste(names(brainflowprobes::pd), collapse = "', '"), "'.",
                call. = FALSE)
        }
        ## Keep the main ones (in case the user shuffled them)
        COVERAGE <- COVERAGE[names(brainflowprobes::pd)]
        if(!all(vapply(COVERAGE, is.list, logical(1)))) {
            stop("Each of the elements of 'COVERAGE' should be a list of\n",
                "region coverage data.frames as created by\n",
                "brainflowprobes_cov().",
                call. = FALSE)
        }
        if(!all(
            vapply(COVERAGE, function(x) is.data.frame(x[[1]]), logical(1)))) {
            stop("Each of the elements of 'COVERAGE' should be a list of\n",
                "region coverage data.frames as created by\n",
                "brainflowprobes_cov().",
                call. = FALSE)
        }
        if(!identical(
            vapply(brainflowprobes::pd, nrow, integer(1)),
            vapply(COVERAGE, function(x) ncol(x[[1]]), integer(1))
        )) {
            stop("Each of the region coverage data.frame lists in 'COVERAGE'\n",
                "should have a column per each of the samples in\n",
                "brainflowprobes::pd.", call. = FALSE)
        }
        if(length(unique(
            vapply(COVERAGE, function(x) nrow(x[[1]]), integer(1)))) != 1) {
            stop("Each of the region coverage data.frames inside 'COVERAGE'\n",
                "should have the same number of rows (1 row per base-pair).",
                call. = FALSE)
        }
        regionCov <- COVERAGE
    }
    return(regionCov)
}


R/region_info.R
#' Print relevant info about candidate probe sequence.
#'
#' `region_info` returns annotation of a single potential probe sequence or
#' list of sequences and, if specified, prints the resuts in a .csv file.
#'
#' @param REGION Either a single hg19 genomic sequence including the chromosome,
#'   start, end, and optionally strand separated by colons (e.g.,
#'   'chr20:10199446-10288068:+'), or a string of sequences to be annotated.
#'   Must be character. Chromosome must be proceeded by 'chr'.
#' @param CSV A logical value indicating if the results should be exported in a
#'   .csv file.
#' @param SEQ A logical value indicating if the base sequence should be
#'   returned.
#' @param PATH If a .csv file is to be exported, this parameter indicates the
#'   path where the file should be saved. By default the file will be
#'   saved in the working directory.
#' @inheritParams four_panels
#' @return This function annotates all input sequences using
#'   [bumphunter::matchGenes()]. It returns a data frame where each
#'   row is a genomic sequence specified in REGION. The columns
#'   c('seqnames', 'start', 'end', 'width', 'strand') list the chromosome,
#'   range, sequence length, and strand of the REGION. The columns c('name',
#'   'annotation', 'description', 'region', 'distance', 'subregion',
#'   'insideDistance', 'exonnumber', 'nexons', 'UTR', 'geneL', 'codingL',
#'   'Geneid', 'subjectHits') are described in
#'   [bumphunter::matchGenes()] documentation.
#'
#'   If SEQ=TRUE, a column 'Sequence' will be included. This is recommended for
#'   sending the probe sequence to be synthesized.
#'
#'   If CSV=TRUE, a .csv file called region_info.csv will be saved to the
#'   working directory unless otherwise specified in PATH.
#' @examples
#' x <- region_info('chr20:10286777-10288069:+', CSV = FALSE)
#' head(x)
#'
#' ## You can easily transform this data.frame to a GRanges object
#' GenomicRanges::GRanges(x)
#'
#' y <- region_info(c('chr20:10286777-10288069:+',
#'                    'chr18:74690788-74692427:-',
#'                    'chr19:49932861-49933829:-'),
#'                  CSV = FALSE, SEQ = FALSE)
#' head(y)
#'
#' candidates <- c('chr20:10286777-10288069:+',
#'                 'chr18:74690788-74692427:-',
#'                 'chr19:49932861-49933829:-')
#' region_info(candidates, CSV = FALSE)
#'
#' ## Explore the effect of changing CODING_ONLY
#' ## Check how the "distance", "name", "Geneid" among other values change
#' region_info('chr10:135379301-135379311:+', CSV = FALSE)
#' region_info('chr10:135379301-135379311:+', CSV = FALSE, CODING_ONLY = TRUE)
#'
#'
#' \dontrun{
#' region_info(candidates, PATH = '/path/to/directory/')
#'
#' region_info('chr20:10286777-10288069:+', PATH = '/path/to/directory')
#'
#' }
#' @export
#' @import GenomicRanges bumphunter Biostrings BSgenome.Hsapiens.UCSC.hg19
#' @importFrom utils write.csv
#' @author Amanda J Price


region_info <- function(REGION, CSV = TRUE, SEQ = TRUE, PATH = ".",
    CODING_ONLY = FALSE) {

    ## Define the region(s)
    gr <- GenomicRanges::GRanges(REGION)

    ## Compute the nearest annotation
    nearestAnnotation <- get_nearest_annotation(gr, CODING_ONLY)
    nearestAnnotation <- nearestAnnotation[,
        -which(colnames(nearestAnnotation) %in%
            c("strand", "subjectHits"))]

    if (SEQ) {
        df <- cbind(as.data.frame(gr),
            nearestAnnotation,
            Sequence = as.character(Biostrings::getSeq(
                BSgenome.Hsapiens.UCSC.hg19::Hsapiens,
                gr)))
    } else {
        df <- cbind(as.data.frame(gr),
            nearestAnnotation)
    }

    if (CSV) {
        csv_path <- file.path(PATH,
            "region_info.csv")
        if (file.exists(csv_path))
            stop(paste("The file",
                csv_path,
                "already exists! Rename or erase it before proceeding."))
        utils::write.csv(df,
            file = csv_path,
            quote = FALSE,
            row.names = FALSE)

        if (!file.exists(csv_path)) {
            stop(paste("Check that the specified path exists.",
                "(Are you missing a backslash?)"))
        }
    }

    message("Completed! If CSV=TRUE, check for region_info.csv in your working
        directory unless otherwise specified in PATH.")
    return(df)

}


R/plot_coverage.R
#' Plot coverage in nuclear and cytoplasmic RNA for candidate probe sequences.
#'
#' `plot_coverage` plots the coverage of nuclear and cytoplasmic RNA from
#' adult and prenatal human prefrontal cortex across a candidate or series of
#' candidate probe sequences for BrainFlow. If the sequence spans splice
#' junctions, the plot will include the introns. A good candidate sequence will
#' be highly and evenly expressed in nuclear RNA.
#'
#' @param PDF The path and name of the PDF file. Defaults to
#' `regionCoverage_fractionedData.pdf`.
#' @inheritParams four_panels
#' @return `plot_coverage` plots all input sequences using
#'   [derfinderPlot::plotRegionCoverage()]. It returns a plot for each
#'   input candidate sequence listed in REGION. Each plot includes coverage of
#'   the sequence(s) in nuclear (N) and cytoplasmic (C) RNA isolated from adult
#'   (A) and fetal (F) prefrontal cortex, sequenced using two different library
#'   preparation methods. PolyA+ libraries (P) were generated using selection
#'   for polyadenylated transcripts, and RiboZero (R) libraries were generated
#'   using a ribosomal RNA depletion protocol.
#'
#'   Each plot also shows the overlapping genes beneath the coverage, and the
#'   genomic location. The title lists the nearest gene, the position of the
#'   sequence relative to the gene's canonical transcriptional start site (TSS),
#'   and further annotation information as described in the 'region' column from
#'   [bumphunter::matchGenes()].
#'
#'   `plot_coverage` saves the results as regionCoverage_fractionedData.pdf
#'   in the working directory unless otherwise specified in PATH.
#' @examples
#'
#' ## Here we use the pre-saved example coverage data such that this example
#' ## will run fast!
#' plot_coverage('chr20:10286777-10288069:+',
#'     COVERAGE = four_panels_example_cov)
#'
#' ## Without using COVERAGE, this function reads BigWig files from the web
#' ## using rtracklayer and this functionality is not supported on Windows
#' ## machines.
#' if(.Platform$OS.type != 'windows') {
#'     plot_coverage('chr20:10286777-10288069:+',
#'         PDF = 'regionCoverage_fractionedData_fromScratch.pdf')
#' }
#'
#' \dontrun{
#'
#' ## These examples will take a few minutes to run!
#' plot_coverage(c('chr20:10286777-10288069:+',
#'                 'chr18:74690788-74692427:-',
#'                 'chr19:49932861-49933829:-'))
#'
#' candidates <- c('chr20:10286777-10288069:+',
#'                 'chr18:74690788-74692427:-',
#'                 'chr19:49932861-49933829:-')
#'
#' ## General syntax:
#' plot_coverage(candidates, PDF = '/path/to/directory/PDF_file.pdf')
#'
#' plot_coverage('chr20:10286777-10288069:+',
#'     PDF = '/path/to/directory/PDF_file.pdf')
#'
#' ## Explore the effect of changing CODING_ONLY
#' ## Check how gene name and distance to TSS changes in the title of the plot
#' ## (everything else stays the same)
#' cov <- brainflowprobes_cov('chr10:135379301-135379311:+')
#' plot_coverage('chr10:135379301-135379311:+', COVERAGE = cov)
#' plot_coverage('chr10:135379301-135379311:+', COVERAGE = cov,
#'     PDF = 'coding_only_plot_coverage', CODING_ONLY = TRUE)
#' }
#'
#' @export
#' @import GenomicRanges bumphunter derfinder derfinderPlot RColorBrewer
#' @importFrom utils browseURL
#' @importFrom grDevices pdf dev.off
#' @author Amanda J Price


plot_coverage <- function(REGION,
    PDF = "regionCoverage_fractionedData.pdf",
    COVERAGE = NULL,
    CODING_ONLY = FALSE,
    VERBOSE = TRUE) {

    ## Check the PDF file
    pdf_file <- check_pdf(PDF)

    ## Define the region(s)
    gr <- GenomicRanges::GRanges(REGION)

    ## Compute the nearest annotation
    nearestAnnotation <- get_nearest_annotation(gr, CODING_ONLY)

    ## Annotate the regions
    annotatedRegions <- derfinder::annotateRegions(regions = gr,
        genomicState = brainflowprobes::gs,
        minoverlap = 1)

    ## Compute or check the coverage (only for Sep)
    regionCov <- get_region_cov(REGION, COVERAGE, VERBOSE,
        PD = brainflowprobes::pd['Sep'])$Sep

    grDevices::pdf(pdf_file, height = 8, width = 8, useDingbats = FALSE)

    derfinderPlot::plotRegionCoverage(regions = gr,
        regionCoverage = regionCov,
        groupInfo = brainflowprobes::pd$Sep$Shortlabels,
        colors = RColorBrewer::brewer.pal(8, "Paired"),
        nearestAnnotation = nearestAnnotation,
        annotatedRegions = annotatedRegions,
        whichRegions = seq_len(length(gr)),
        ask = FALSE,
        verbose = FALSE
    )
    grDevices::dev.off()

    .view_pdf(pdf_file)
    return(invisible(pdf_file))
}

.view_pdf <- function(pdf_file) {
    message(paste(Sys.time(), "Completed! Check for ",
        pdf_file, " in your working directory unless otherwise specified."))
    if (interactive())
        utils::browseURL(pdf_file)
}


R/four_panels.R
#' Plot expression coverage in four datasets for candidate probe sequences.
#'
#' `four_panels` creates four plots for each candidate probe sequence. The
#' first plot (Separation) shows the adjusted read coverage in cytosolic and
#' nuclear RNA from human postmortem cortex. The second plot (Degradation) shows
#' coverage in human cortical samples exposed to room temperature for 0-60
#' minutes. The third plot (Sorted) shows RNA coverage in nuclei that had been
#' sorted based on reactivity to NeuN-antibody, a neuronal marker. NeuN+ samples
#' are enriched for neurons, and NeuN- samples are enriched for non-neurons. The
#' fourth plot (Single Cells) shows the expression coverage in single cells
#' isolated from human temporal lobe.
#'
#' @param REGION Either a single hg19 genomic sequence including the chromosome,
#'   start, end, and optionally strand separated by colons (e.g.,
#'   'chr20:10199446-10288068:+'), or a string of sequences. Must be character.
#'   Chromosome must be proceeded by 'chr'.
#' @param PDF The path and name of the PDF file. Defaults to
#' `four_panels.pdf`.
#' @param JUNCTIONS A logical value indicating if the candidate probe sequence
#'   spans splice junctions (Default=FALSE).
#' @param COVERAGE The output of [brainflowprobes_cov] for the input
#' `REGION`. Defaults to `NULL` but it can be pre-computed and saved
#' separately since this is the step that takes the longest to run. Also, this
#' is the only step that depends on rtracklayer's functionality for reading
#' BigWig files which does not run on Windows OS. So it could be run on a
#' non-Windows machine, saved, and then shared with Windows users.
#' @param CODING_ONLY A logical vector of length 1 specifying whether to
#' subset [genes] to only the coding genes. That is, whether to subset
#' [genes] by whether they have a non-NA `CSS` value.
#' @param VERBOSE A logical value indicating whether to print updates from the
#' process of loading the data from the BigWig files.
#' @return `four_panels()` first annotates the input candidate probe
#'   sequence(s) in REGION using [bumphunter::matchGenes()], and then
#'   cuts the expression coverage for each sequence from each sample in four
#'   different datasets (see the BrainFlow publication for references) using
#'   [derfinder::getRegionCoverage()]. The coverage is normalized to
#'   the total mapped reads per sample and kilobase width of each probe region
#'   before log2 transformation. The four plots are labeled by the dataset and
#'   the plots are topped by the sequence coordinates, sequence width, and the
#'   name of the nearest gene.
#'
#'   A good candidate probe sequence will have several characteristics. In the
#'   Separation data, the sequence should be relatively highly expressed in
#'   nuclear RNA, at least in your age of interest. The sequence should also
#'   show stable expression over the 60 minutes of room temperature exposure in
#'   the Degradation data. The sequence should also be expressed in the
#'   appropriate NeuN fraction (depending on cell type specificity) in the
#'   Sorted dataset, and also be expressed in the right cell type in the Single
#'   Cell dataset.
#'
#'   `four_panels()` saves the results as four_panels.pdf in the working
#'   directory unless otherwise specified in PATH.
#'
#'   `if(JUNCTIONS)`, this means that the candidate probe sequence spans
#'   splice junctions. In this case, the character vector of regions should
#'   represent the coordinates of each exon spanned in the sequence.
#'   `if(JUNCTIONS)`, `four_panels()` will sum the coverage of each exon and
#'   plot that value for each dataset instead of creating an independent set of
#'   plots for each exon. This is a way to avoid deflating coverage by including
#'   lowly-expressed intron coverage in the plots.
#' @examples
#'
#' ## Here we use the pre-saved example coverage data such that this example
#' ## will run fast!
#' four_panels('chr20:10286777-10288069:+',
#'     COVERAGE = four_panels_example_cov)
#'
#' \dontrun{
#' ## Without using COVERAGE, this function reads BigWig files from the web
#' ## using rtracklayer and this functionality is not supported on Windows
#' ## machines.
#' if(.Platform$OS.type != 'windows') {
#'     ## This example takes 10 minutes to run!
#'     four_panels('chr20:10286777-10288069:+')
#' }
#'
#' ## These examples will take several minutes to run depending on your
#' ## internet connection
#' four_panels(c(
#'     'chr20:10286777-10288069:+',
#'     'chr18:74690788-74692427:-',
#'     'chr19:49932861-49933829:-'))
#'
#' PENK_exons <- c(
#'     'chr8:57353587-57354496:-',
#'     'chr8:57358375-57358515:-',
#'     'chr8:57358985-57359040:-',
#'     'chr8:57359128-57359292:-')
#'
#' ## General syntax
#' four_panels(PENK_exons, JUNCTIONS=TRUE,
#'     PDF = '/path/to/directory/PDF_file.pdf')
#'
#' four_panels('chr20:10286777-10288069:+',
#'     PDF = '/path/to/directory/PDF_file.pdf')
#'
#'
#' ## Explore the effect of changing CODING_ONLY
#' ## Check how gene name changes in the title of the plot
#' ## (everything else stays the same)
#' cov <- brainflowprobes_cov('chr10:135379301-135379311:+')
#' four_panels('chr10:135379301-135379311:+', COVERAGE = cov)
#' four_panels('chr10:135379301-135379311:+', COVERAGE = cov,
#'     PDF = 'coding_only_four_panels', CODING_ONLY = TRUE)
#' }
#' @export
#' @import GenomicRanges bumphunter ggplot2 derfinder RColorBrewer cowplot
#' @importFrom utils browseURL
#' @importFrom grDevices pdf dev.off
#' @author Amanda J Price


four_panels <- function(REGION,
    PDF = "four_panels.pdf",
    JUNCTIONS = FALSE,
    COVERAGE = NULL,
    CODING_ONLY = FALSE,
    VERBOSE = TRUE) {


    ## For R CMD check
    BrNum <- Cell_type <- Cov <- DegradationTime <- Label <- LabelFrac <-
        LibraryProtocol <- Shortlabels <- NULL

    ## Check the PDF file
    pdf_file <- check_pdf(PDF)

    ## Define the region(s)
    gr <- GenomicRanges::GRanges(REGION)

    ## Compute the nearest annotation
    nearestAnnotation <- get_nearest_annotation(gr, CODING_ONLY)

    ## Compute or check the coverage
    regionCov <- get_region_cov(REGION, COVERAGE, VERBOSE)

    if(JUNCTIONS) {
        regionCov <- lapply(regionCov, function(x) list(do.call(rbind, x)))
        coords <- paste0(GenomicRanges::seqnames(gr)[1],
            ":", min(GenomicRanges::start(gr)),
            "-", max(GenomicRanges::end(gr)),
            ":", GenomicRanges::strand(gr)[1])

    } else {
        coords <- paste0(GenomicRanges::seqnames(gr),
            ":", GenomicRanges::start(gr),
            "-", GenomicRanges::end(gr),
            ":", GenomicRanges::strand(gr))
    }

    covMat <- lapply(regionCov, function(x) t(sapply(x, colSums)/100))

    bg <- lapply(covMat, function(x) matrix(sum(GenomicRanges::width(gr)),
            ncol = ncol(x),
            nrow = nrow(x))/1000)

    mains <- paste0(coords,
        " (", sum(GenomicRanges::width(gr)),
        " bp)\n",
        nearestAnnotation$name)


    covMat <- mapply(function(cv, b) cv/b, covMat, bg, SIMPLIFY = FALSE)
    covMat <- lapply(covMat, function(x) as.matrix(log2(x + 1)))
    theRanges <- range(do.call(cbind, covMat))

    grDevices::pdf(pdf_file, height = 12, width = 10, useDingbats = FALSE)

    result <- vector(mode = "list", length = length(regionCov[[1]]))
    for (j in seq_len(length(regionCov[[1]]))) {

        SepDat <- data.frame(Cov = covMat$Sep[j,
            ], LabelFrac = brainflowprobes::pd$Sep$LabelFrac,
            Shortlabels = brainflowprobes::pd$Sep$Shortlabels)
        Sep <- ggplot2::ggplot(SepDat,
            aes(x = LabelFrac, y = Cov)) +
            theme_bw() +
            geom_boxplot(outlier.shape = NA) +
            geom_jitter(size = 2,
                aes(fill = Shortlabels),
                pch = 21,
                color = "black") +
            scale_fill_manual(values = RColorBrewer::brewer.pal(8,
                "Paired")) +
            labs(fill = "") +
            ylim(theRanges) +
            ylab("Log2(Adj Read/kb)") +
            xlab("") +
            theme(text = element_text(size = 20),
                legend.text = element_text(size = 12),
                legend.position = c(0.5,
                    0.15),
                legend.background = element_rect(fill = "transparent",
                    linetype = "solid",
                    colour = "black"),
                legend.title = element_blank(),
                axis.text.x = element_text(angle = 90,
                    hjust = 1)) +
            guides(fill = guide_legend(ncol = 4)) +
            ggtitle("Separation")


        DegDat <- data.frame(Cov = covMat$Deg[j,
            ], DegradationTime = brainflowprobes::pd$Deg$DegradationTime,
            LibraryProtocol = brainflowprobes::pd$Deg$LibraryProtocol,
            BrNum = brainflowprobes::pd$Deg$BrNum)
        Deg <- ggplot2::ggplot(DegDat,
            aes(x = DegradationTime,
                y = Cov,
                fill = BrNum,
                color = BrNum)) +
            theme_bw() +
            geom_line(aes(linetype = LibraryProtocol),
                lwd = 1.3) +
            geom_point(cex = 2,
                pch = 21,
                aes(fill = BrNum),
                color = "black") +
            scale_color_manual(values = RColorBrewer::brewer.pal(5,
                "Dark2")) +
            scale_fill_manual(values = RColorBrewer::brewer.pal(5,
                "Dark2")) +
            ylim(theRanges) +
            ylab("") +
            xlab("") +
            theme(text = element_text(size = 20),
                legend.text = element_text(size = 12),
                legend.position = c(0.5,
                    0.1),
                legend.background = element_rect(fill = "transparent",
                    linetype = "solid",
                    colour = "black"),
                legend.title = element_blank()) +
            guides(linetype = guide_legend(ncol = 2),
                fill = FALSE,
                color = FALSE) +
            ggtitle("Degradation")


        SortDat <- data.frame(Cov = covMat$Sort[j,
            ], Label = gsub(":",
            "\n", levels(factor(brainflowprobes::pd$Sort$Label))))
        Sort <- ggplot2::ggplot(SortDat,
            aes(x = Label,
                y = Cov)) +
            theme_bw() +
            geom_boxplot(outlier.shape = NA) +
            geom_jitter(size = 2,
                aes(fill = Label),
                pch = 21,
                color = "black") +
            scale_fill_manual(values = unique(brainflowprobes::pd$Sort$col)) +
            ylim(theRanges) +
            ylab("Log2(Adj Read/kb)") +
            xlab("") +
            theme(text = element_text(size = 20),
                axis.text.x = element_text(angle = 90,
                    vjust = 0.5,
                    hjust = 1)) +
            guides(fill = FALSE) +
            ggtitle("Sorted")


        CellDat <- data.frame(Cov = covMat$Cell[j,
            ], Cell_type = brainflowprobes::pd$Cell$Cell_type)
        Cell <- ggplot2::ggplot(CellDat,
            aes(x = Cell_type,
                y = Cov)) +
            theme_bw() +
            geom_boxplot(outlier.shape = NA) +
            geom_jitter(aes(fill = Cell_type),
                pch = 21,
                color = "black") +
            scale_fill_manual(values = c(
                "#FFBB78",
                "#FF9896",
                "#C5B0D5",
                "#9467BD",
                "#1F77B4",
                "#2CA02C",
                "#D62728",
                "#FF7F0E",
                "#AEC7E8",
                "#98DF8A")) +
            ylim(theRanges) +
            ylab("") +
            xlab("") +
            theme(text = element_text(size = 20),
                axis.text.x = element_text(angle = 90,
                    vjust = 0.5,
                    hjust = 1)) +
            guides(fill = FALSE) +
            ggtitle("Single Cells")

        p <- cowplot::plot_grid(Sep, Deg, Sort, Cell, ncol = 2, align = "hv")
        title <- cowplot::ggdraw() +
            cowplot::draw_label(mains[j],
                fontface = "bold",
                size = 22)
        g <- cowplot::plot_grid(title,
            p, ncol = 1,
            rel_heights = c(0.1,
                1))
        print(g)

        result[[j]] <- g
    }
    grDevices::dev.off()

    .view_pdf(pdf_file)
    return(invisible(j))

}
